src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs

```diff
@@ -0,0 +1,3664 @@
+}
+            ";
+            return VerifyCS.VerifyCodeFixAsync(originalCode, fixedCode);
+        }
+
+        [Fact]
+        public Task CS_Diagnostic_Default_WithNamedParametersUnordered()
+        {
+            string originalCode = @"
+using System.Threading;
+class C
+{
+    int M(CancellationToken ct)
+    {
+        return [|MyMethod|](z: ""Hello world"", x: 5, y: true);
+    }
+    int MyMethod(int x, bool y = default, string z = """", CancellationToken c = default) => 1;
+}
+            ";
+            // Notice the parameters do NOT get reordered to their official position
+            string fixedCode = @"
+using System.Threading;
+class C
+{
+    int M(CancellationToken ct)
+    {
+        return MyMethod(z: ""Hello world"", x: 5, y: true, c: ct);
+    }
+    int MyMethod(int x, bool y = default, string z = """", CancellationToken c = default) => 1;
+}
+            ";
+            return VerifyCS.VerifyCodeFixAsync(originalCode, fixedCode);
+        }
+
+        [Fact]
+        public Task CS_Diagnostic_WithLock()
+        {
+            string originalCode = @"
+using System.Threading;
+class C
+{
+    private readonly object lockingObject = new object();
+    int M (CancellationToken ct)
+    {
+        int x;
+        lock (lockingObject)
+        {
+            x = [|MyMethod|](5);
+        }
+        return x;
+    }
+    int MyMethod(int x, CancellationToken c = default) => 1;
+}
+            ";
+            string fixedCode = @"
+using System.Threading;
+class C
+{
+    private readonly object lockingObject = new object();
+    int M (CancellationToken ct)
+    {
+        int x;
+        lock (lockingObject)
+        {
+            x = MyMethod(5, ct);
+        }
+        return x;
+    }
+    int MyMethod(int x, CancellationToken c = default) => 1;
+}
+            ";
+            return VerifyCS.VerifyCodeFixAsync(originalCode, fixedCode);
+        }
+
+        [Fact]
+        public Task CS_Diagnostic_DereferencePossibleNullReference()
+        {
+            string originalCode = @"
+#nullable enable
+using System.Threading;
+class C
+{
+    O? PossiblyNull()
+    {
+        return null;
+    }
+    void M(CancellationToken ct)
+    {
+        O? o = PossiblyNull();
+        o?.[|MyMethod|]();
+    }
+}
+class O
+{
+    public int MyMethod(CancellationToken c = default) => 1;
+}
+            ";
+            string fixedCode = @"
+#nullable enable
+using System.Threading;
+class C
+{
+    O? PossiblyNull()
+    {
+        return null;
+    }
+    void M(CancellationToken ct)
+    {
+        O? o = PossiblyNull();
+        o?.MyMethod(ct);
+    }
+}
+class O
+{
+    public int MyMethod(CancellationToken c = default) => 1;
+}
+            ";
+            // Nullability is available in C# 8.0+
+            return CSharp8VerifyCodeFixAsync(originalCode, fixedCode);
+        }
+
+        [Fact]
+        public Task CS_Diagnostic_LambdaAndExtensionMethod()
```

Comment:
@mavasani this is the method that is currently failing because I am checking the `ContainingSymbol` for a CancellationToken among the arguments, instead of navigating through the parents and finding the very first container.

`ContainingSymbol` in this example will return `M(CancellationToken ct)`, when in fact the container is `b.Extension((j)`.

If I could use the semantic model to check each ancestor, I would be able to get this case addressed. Or if you know an alternative similar to the suggestion you gave me to use `ContainingSymbol`, without having to use the semantic model, that would be great.

I need to address this case because I don't have an (easy) way to detect the type of the parameter in the lambda, and the fixer is assuming that's the token, and that's the argument that it is passing to `MyMethod`.

Reply:
You should walk up the operation tree until you either see an IAnonymousFunctionOperation or an ILocalFunctionOperation or null. First two operation nodes expose the corresponding lambda or local function method symbol respectively, and you should default to ContainingSymbol for the null case. It would be good to write this as an extension method in IOperationExtensions.cs.

Reply:
Actually, I am wondering if we should loop through all the containing lambda, local functions and finally top level containing method and if any of them have a cancellation token parameter, report the diagnostic. If so, you will basically call this extension method in a loop and validate.

Reply:
Let me write a unit test with a bunch of different nested methods. The code, as it is right now, should be able to find a cancellationToken eventually (because it looks for all ancestors until null).

Reply:
@mavasani I can't get this addressed easily because I am only able to detect the type of the parameter when I have access to the operations, which means I can detect any ancestor in the analyzer if it passes a token. But in the fixer, since I only have access to the syntax, I am unable to know the type of the parameter, and this is particularly problematic with lambdas/anonymous functions.

So I would like to keep this version simple, and only trigger a diagnostic if the immediate ancestor is the one with the token, so that the fixer doesn't have to try to find which ancestor had the actual token, and simply grab the last parameter from the immediate ancestor.

Reply:
Fixer does not need to offer a fix for all cases - it can only a fix for the case where the cancellation token parameter is in the immediate enclosing function. I think that should not prevent the analyzer from flagging that case though. As a general rule, analyzer implementation should never be influenced by fixer support.

Reply:
You should be able to a simple property to the diagnostic indicating if the cancellation parameter is in immediate enclosing method or not, and fixer just bails out for diagnostics where cancellation token is not in immediate method.

Reply:
Interesting. I didn't know diagnostics could have additional properties. Can you please show me an example?

Reply:
Yep, https://github.com/dotnet/roslyn-analyzers/blob/0743db3bd6950343d127b26b40b303b3fb7b0bfb/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs#L369-L379